tidy3d/components/geometry/utils.py

```diff
@@ -206,3 +212,128 @@ def validate_no_transformed_polyslabs(geometry: GeometryType, transform: MatrixR
     elif isinstance(geometry, base.ClipOperation):
         validate_no_transformed_polyslabs(geometry.geometry_a, transform)
         validate_no_transformed_polyslabs(geometry.geometry_b, transform)
+
+
+class SnapLocation(Enum):
+    """Describes different methods for defining the snapping locations."""
+
+    Boundary = 1
+    """
+    Choose the boundaries of Yee cells.
+    """
+    Center = 2
+    """
+    Choose the center of Yee cells.
+    """
+
+
+class SnapBehavior(Enum):
+    """Describes different methods for snapping intervals, which are defined by two endpoints."""
+
+    Closest = 1
+    """
+    Snaps the interval's endpoints to the closest grid point.
+    """
+    Expand = 2
+    """
+    Snaps the interval's endpoints to the closest grid points,
+    while guaranteeing that the snapping location will never move endpoints inwards.
+    """
+    Contract = 3
+    """
+    Snaps the interval's endpoints to the closest grid points,
+    while guaranteeing that the snapping location will never move endpoints outwards.
+    """
+    Off = 4
+    """
+    Do not use snapping.
+    """
+
+
+class SnappingSpec(Tidy3dBaseModel):
+    """Specifies how to apply grid snapping along each dimension."""
+
+    location: tuple[SnapLocation, SnapLocation, SnapLocation] = pydantic.Field(
+        ...,
+        title="Location",
+        description="Describes which positions in the grid will be considered for snapping.",
+    )
+
+    behavior: tuple[SnapBehavior, SnapBehavior, SnapBehavior] = pydantic.Field(
+        ...,
+        title="Behavior",
+        description="Describes how snapping positions will be chosen.",
+    )
+
+
+def snap_box_to_grid(grid: Grid, box: Box, snap_spec: SnappingSpec, rtol=fp_eps) -> Box:
+    """Snaps a :class:`.Box` to the grid, so that the boundaries of the box are aligned with grid centers or boundaries.
+    The way in which each dimension of the `box` is snapped to the grid is controlled by ``snap_spec``.
+    """
+
+    def closest_value(test: float, first: float, second: float) -> float:
+        """Helper to choose the closest value to a given test value."""
+        dfirst = np.abs(test - first)
+        dsecond = np.abs(test - second)
```

Comment:
`abs` instead of `np.abs` for performance gain: `np.abs` will construct a numpy array.